lib/ansible/modules/windows/win_msp.py

```diff
@@ -0,0 +1,152 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# Copyright: (c) 2019, Rodric Vos <rodric@vosenterprises.eu>, and others
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+# this is a windows documentation stub. Actual code lives in the .ps1
+# file of the same name.
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = r'''
+---
+module: win_msp
+short_description: Installs/uninstalls a Microsoft Installer Patch file (.msp).
+description:
+- Installs or uninstalls a package in MSP format.
+- These packages can be sources from the local file system, network file share
+  or a url.
+version_added: "2.8"
+author: Rodric Vos (@finzzownt)
+options:
+  expected_return_code:
+    description:
+    - One or more return codes from the package installation that indicates
+      success.
+    - A return code of C(3010) usually means that a reboot is required, the
+      I(reboot_required) return value is set if the return code is C(3010).
+    type: list
+    default: [0, 3010]
+  path:
+    description:
+    - Location of the MSP to be installed, e.g. I(state=present).
+    - This package can either be on the local file system, network share or a
```

Comment:
Implementation of downloading a path file - especially from http(s) - shouldn't be, IMHO, the part of that module. Please use [win_get_url](https://docs.ansible.com/ansible/latest/modules/win_get_url_module.html#win-get-url-module) instead.

Reply:
I agree totally. For instance I've extended the win_get_url module to support downloading and checksum validation. One would argue why not extend that functionality in win_msp too? This was a refactor of the previous win_msi module. It's active in a production environment for six months now, and only now I found the time to contribute. By now it's outdated I noticed. Okay, long story short: I'll remove http(s) support and refer to win_get_url in the documentation.